src/modules/anchor.js

```diff
@@ -0,0 +1,119 @@
+  /**
+   * Creates a new instance of AnchorModule and sets the api
+   * @constructor
+   * @param {object} api - PolkadotJS API Reference
+   * @param {Function} signAndSend - Callback signing and sending
+   */
+  constructor(api, signAndSend) {
+    this.api = api;
+    this.module = api.tx.anchor;
+    this.signAndSend = signAndSend;
+  }
+
+  /**
+   * Prepare transaction to write anchor on chain
+   * @param {Uint8Array|string} anchor
+   * @return {object} The extrinsic to sign and send.
+   */
+  deployTx(anchor) {
+    const toPost = normalizeToHex(anchor);
+    return this.module.deploy(toPost);
+  }
+
+  /**
+   * Write anchor on chain
+   * @param anchor
+   * @param waitForFinalization
+   * @param params
+   * @returns {Promise<*>}
+   */
+  async deploy(anchor, waitForFinalization = true, params = {}) {
+    return await this.signAndSend(this.deployTx(anchor), waitForFinalization, params);
+  }
+
+  /**
+   * Query anchor from chain
+   * @param anchor
+   * @param {Boolean} preHashed - If the anchor has already been hashed.
+   * @returns {Promise<*>} - The promise will either successfully resolve to the block number where anchor was created
+   * or reject with an error.
+   */
+  async get(anchor, preHashed = false) {
+    let key;
+    if (preHashed) {
+      key = normalizeToHex(anchor);
+    } else if (anchor instanceof Uint8Array) {
+      key = u8aToHex(this.hash(anchor));
+    } else if (isHexWithGivenByteSize(anchor)) {
+      key = u8aToHex(this.hash(hexToU8a(anchor)));
+    } else {
+      throw new Error('Require a hex string or a byte array');
+    }
+
+    const resp = await this.api.query.anchor.anchors(key);
+    if (resp.isNone) {
+      throw new NoAnchorError(anchor);
+    }
+    return resp.unwrap().toNumber();
+  }
+
+  /**
+   * Batch multiple documents in binary merkle tree and return the root and proofs for each document
+   * @param documents
+   * @returns {Array} - An 2 element array where 1st element is root and 2nd is an array with proofs for
+   * each document.
+   */
+  batchDocumentsInMerkleTree(documents) {
+    // Hash all documents
+    const leafHashes = documents.map(this.hash);
+    // Concatenate all leaf hashes into one bytearray
+    const packed = new Uint8Array(leafHashes.map((a) => [...a]).flat());
+    const [root, proofs] = construct(packed);
+    return [Uint8Array.from(root), proofs];
+  }
+
+  /**
+   * Verify inclusion proof of document in a merkle tree with given root. The document is hashed to form a leaf first
+   * @param document
+   * @param proof
+   * @param root
+   * @returns {boolean}
+   */
+  verifyMerkleProofOfDocument(document, proof, root) {
+    const hash = this.hash(document);
+    return this.verifyMerkleProofOfLeaf(hash, proof, root);
+  }
+
+  /**
+   * Verify inclusion proof of leaf in a merkle tree with given root.
+   * @param leaf
+   * @param proof
+   * @param root
+   * @returns {boolean}
+   */
+  verifyMerkleProofOfLeaf(leaf, proof, root) {
```

Comment:
What is the suggestion? Are you saying root should be optional? For the verification to complete, root must be known

Reply:
Root can be inferred if the leaf and proof are known.

Just mentioning it because there are use case where the leaf and the proof are known but not the root. That's why [verify_proof was changed to return the root rather than take it as an argument](https://github.com/docknetwork/mrklt/commit/96d1eaaf2088804006c26923a849186e27deee6e#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759L39).

I don't see a problem leaving `verifyMerkleProofOfLeaf` as-is; the use case is still covered because the user can call `verify_proof` directly when they need to infer the root. If you are ok with that I suggest marking this conversation as resolved.

Reply:
When would you want to infer the root?

The verifier should never try to infer the root from the proof and leaf. He must know the root.

Reply:
Usually for Merkle trees that would be the case. Checking against a specific root is the whole point. In the context of batched anchors the user doesn't care whether the root is equal to some root that was provided; in fact the root doesn't even need to be provided. What they care about is when it was posted.

```rust
// Note inferred roots are already proven to be valid roots.
let root = verify_proof(leaf, proof);
assert_eq!(root, verify_proof(leaf, proof));
```

Reply:
If a user (not the verifier) does not care about root, then he is not verifying but computing the root and that logic should be in another function and not in verifcation.